bpf/dns_tracker.h

```diff
@@ -0,0 +1,141 @@
+        }
+    }
+    return 0;
+}
+
+static inline void find_or_create_dns_flow(flow_id *id, struct dns_header *dns, int len, int dir, u16 flags) {
+    id->direction = dir;
+    flow_metrics *aggregate_flow = bpf_map_lookup_elem(&aggregated_flows, id);
+    if (aggregate_flow != NULL) {
+        aggregate_flow->dns_record.id = bpf_ntohs(dns->id);
+        aggregate_flow->dns_record.flags = bpf_ntohs(dns->flags);
+        aggregate_flow->dns_record.req_mono_time_ts = bpf_ktime_get_ns();
+        bpf_map_update_elem(&aggregated_flows, &id, aggregate_flow, BPF_ANY);
+    } else {
+        // there is no matching flows so lets create new one and add the drops
+        u64 current_time = bpf_ktime_get_ns();
+         flow_metrics new_flow;
+        if (dir == INGRESS) {
+            new_flow.start_mono_time_ts = current_time;
+            new_flow.end_mono_time_ts = current_time;
+            new_flow.packets = 1;
+            new_flow.bytes = len;
+            new_flow.flags = flags;
+            new_flow.dns_record.id = bpf_ntohs(dns->id);
+            new_flow.dns_record.flags = bpf_ntohs(dns->flags);
+            new_flow.dns_record.req_mono_time_ts = current_time;
+        } else {
+            new_flow.start_mono_time_ts = current_time;
+            new_flow.end_mono_time_ts = current_time;
+            new_flow.packets = 1;
+            new_flow.bytes = len;
+            new_flow.flags = flags;
+            new_flow.dns_record.id = bpf_ntohs(dns->id);
+            new_flow.dns_record.flags = bpf_ntohs(dns->flags);
+            new_flow.dns_record.rsp_mono_time_ts = current_time;
+        }
+        bpf_map_update_elem(&aggregated_flows, id, &new_flow, BPF_ANY);
+    }
+}
+
+static inline int trace_dns(struct sk_buff *skb) {
+    struct sock *sk = skb->sk;
+    flow_id id;
+    u8 protocol = 0;
+    u16 family = 0,flags = 0, len = 0;
+
+    __builtin_memset(&id, 0, sizeof(id));
+
+    // pull in details from the packet headers and the sock struct
+    bpf_probe_read(&family, sizeof(u16), &sk->__sk_common.skc_family);
+
+    id.if_index = skb->skb_iif;
+
+    // read L2 info
+    set_key_with_l2_info(skb, &id);
+
+    // read L3 info
+    set_key_with_l3_info(skb, family, &id, &protocol);
+
+    switch (protocol) {
+    case IPPROTO_UDP:
+        len = set_key_with_udp_info(skb, &id, IPPROTO_UDP);
+        break;
+    case IPPROTO_TCP:
+        len = set_key_with_tcp_info(skb, &id, IPPROTO_TCP, &flags);
+        break;
+    default:
+        return -1;
+    }
+
+    // check for DNS packets
+    if (id.dst_port == DNS_PORT || id.src_port == DNS_PORT) {
+        struct dns_header dns;
+        bpf_probe_read(&dns, sizeof(dns), (struct dns_header *)(skb->head + skb->transport_header + len));
+
+        struct dns_request_key req_key = {
+            .id = dns.id
+        };
+        struct dns_request_val dns_val;
+
+        __builtin_memset(&dns_val, 0, sizeof(dns_val));
+
+        if ((dns.flags & DNS_QR_FLAG) == 0) { /* dns query */
```

Comment:
See above cf endianness, you should be wrapping the flags value in `bpf_ntohs()`

Reply:
fixed